apps/language_server/test/experimental/provider/code_action/replace_remote_function_test.exs

```diff
@@ -0,0 +1,257 @@
+defmodule ElixirLS.LanguageServer.Experimental.Provider.CodeAction.ReplaceRemoteFunctionTest do
+  alias ElixirLS.LanguageServer.Experimental.CodeMod.Diff
+  alias ElixirLS.LanguageServer.Experimental.Protocol.Requests
+  alias ElixirLS.LanguageServer.Experimental.Protocol.Requests.CodeAction, as: CodeActionRequest
+  alias ElixirLS.LanguageServer.Experimental.Protocol.Types.CodeAction
+  alias ElixirLS.LanguageServer.Experimental.Protocol.Types.CodeAction, as: CodeActionReply
+  alias ElixirLS.LanguageServer.Experimental.Protocol.Types.Diagnostic
+  alias ElixirLS.LanguageServer.Experimental.Protocol.Types.Range
+  alias ElixirLS.LanguageServer.Experimental.Protocol.Types.Position
+  alias ElixirLS.LanguageServer.Experimental.Protocol.Types.TextEdit
+  alias ElixirLS.LanguageServer.Experimental.Provider.CodeAction.ReplaceRemoteFunction
+  alias ElixirLS.LanguageServer.Experimental.SourceFile
+  alias ElixirLS.LanguageServer.Fixtures.LspProtocol
+  alias ElixirLS.LanguageServer.SourceFile.Path, as: SourceFilePath
+
+  import LspProtocol
+  import ReplaceRemoteFunction
+
+  use ExUnit.Case
+  use Patch
+
+  setup do
+    {:ok, _} = start_supervised(SourceFile.Store)
+    :ok
+  end
+
+  defp diagnostic_message(arity) do
+    """
+    Enum.counts/#{arity} is undefined or private. Did you mean:
+
+          * concat/1
+          * concat/2
+          * count/1
+          * count/2
+    """
+  end
+
+  defp code_action(file_body, file_path, line, opts \\ []) do
+    trimmed_body = String.trim(file_body, "\n")
+
+    file_uri = SourceFilePath.to_uri(file_path)
+    SourceFile.Store.open(file_uri, trimmed_body, 0)
+
+    {:ok, range} =
+      build(Range,
+        start: [line: line, character: 0],
+        end: [line: line, character: 0]
+      )
+
+    message =
+      Keyword.get_lazy(opts, :diagnostic_message, fn ->
+        diagnostic_message(1)
+      end)
+
+    diagnostic = Diagnostic.new(range: range, message: message)
+    {:ok, context} = build(CodeAction.Context, diagnostics: [diagnostic])
+
+    {:ok, action} =
+      build(CodeActionRequest,
+        text_document: [uri: file_uri],
+        range: range,
+        context: context
+      )
+
+    {:ok, action} = Requests.to_elixir(action)
+    {file_uri, action}
+  end
+
+  defp assert_expected_text_edits(file_uri, action, expected_name, line) do
+    assert %CodeActionReply{edit: %{changes: %{^file_uri => edits}}} = action
+
+    expected_edits = Diff.diff("counts", expected_name)
+
+    assert edits
+           |> Enum.zip(expected_edits)
+           |> Enum.all?(fn {%TextEdit{new_text: new_text}, %TextEdit{new_text: expected_new_text}} ->
+             new_text == expected_new_text
+           end)
+
+    assert Enum.all?(edits, fn edit -> edit.range.start.line == line end)
+    assert Enum.all?(edits, fn edit -> edit.range.end.line == line end)
+  end
```

Comment:
These kind of asserts are really difficult to reason about. It's hard to see intrinsically if they're correct or not. 
I'm a big fan of behavior based testing, and I think it makes more sense to see that applying the code action to a segment of code achieves the desired result. 

The good news is that `SourceFile` _almost_ has what you need to accomplish this. All that's required is the following function head to be added (provide you alias TextEdit):  
```elixir
  defp apply_change(%__MODULE__{} = source, %TextEdit{} = change) do
    apply_change(source, change.range, change.new_text)
  end
```

and you'll be able to rewrite this in terms of a document transformation rather than comparing text edits, so you could do:
```elixir
expected_doc = """
Enum.count(a)
"""
assert expected_doc == actual_doc
|> code_action("/project/file.ex", 0)
|> apply_selected_action(0)
```
`apply_selected_action` would select the given action from the ones generated, select that index and send the list of `%TextEdit{}`s to `SourceFile.apply_changes`

which is easier to reason about and understand, and should make this test a bit more concise and digestable. 
Also, I'd pull it out later, and make a CodeAction.Case that everyone developing CodeActions could use.